### PR TITLE
CHE-3720: Decouple different agents on different maven modules

### PR DESCRIPTION
**Pull request by tolusha**
### What does this PR do?
Adds module per agent

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/3720

### Previous behavior
All agents are located in the same folder

### Tests written?
Yes

#### Changelog
Decouple different agents on different maven modules

### CODE REVIEW

**Comment by skabashnyuk** [comment]
Can you describe new folders structure?

**Comment by tolusha** [comment]
```
agents
  ├── che-core-api-agent
  ├── che-core-api-agent-shared
  ├── exec
  ├── ls-csharp
  ├── ls-json
  ├── ls-php
  ├── ls-python
  ├── ls-typescript
  ├── unison
  ├── ssh
  │  ├── pom.xml
  │  └── src
  │      └── main
  │         ├── java
  │         │   └── org
  │         │       └── eclipse
  │         │           └── che
  │         │               └── api
  │         │                   └── agent
  │         │                       ├── SshAgent.java
  │         │                       ├── SshAgentLauncher.java
  │         │                       └── SshAgentDescriptorModule.java
  │         └── resources
  │             ├── org.eclipse.che.ssh.json
  │             └── org.eclipse.che.ssh.script.sh
  └── pom.xml
```

**Comment by tolusha** [comment]
also:
```
wsagent
  ...
  ├── che-core-api-debug
  ├── che-core-api-debug-shared
  ├── che-core-api-git
  ├── che-core-api-git-shared
  ├── che-core-api-languageserver
  ├── che-core-api-languageserver-shared
  ├── che-core-api-project
  ├── che-core-api-project-shared
  ├── che-core-git-impl-jgit
  ├── pom.xml
  ├── agent
  │   ├── pom.xml
  │   └── src
  │       ├── main
  │       │   ├── java
  │       │   │   └── org
  │       │   │       └── eclipse
  │       │   │           └── che
  │       │   │               └── api
  │       │   │                   └── agent
  │       │   │                       ├── WsAgent.java
  │       │   │                       ├── WsAgentLauncher.java
  │       │   │                       └── WsAgentDescriptorModule.java
  │       │   └── resources
  │       │       ├── org.eclipse.che.ws-agent.json
  │       │       └── org.eclipse.che.ws-agent.script.sh
  │       └── test
  │           └── java
  │               └── org
  │                   └── eclipse
  │                       └── che
  │                           └── api
  │                               └── agent
  │                                   └── WsAgentLauncherTest.java
  └── wsagent-local

```

**Comment by TylerJewell** [comment]
That documentation helps.  I think @skabashnyuk - at the same time that this is merged, the wiki needs to be updated with an explanation of how this modifies the development workflow.  We have sections at the end of that wiki that explain the repository layout, and they need to be updated with this information.

**Comment by tolusha** [comment]
@TylerJewell 
Does it mean the structure it is ok for you?

**Comment by TylerJewell** [comment]
@tolusha - I think the code structure of `/wsagent` folder is really for @gazarenkov to comment on.  The `/agents` folder is ok for me.

File: agents/che-core-api-agent/src/main/java/org/eclipse/che/api/agent/server/impl/AgentRegistryImpl.java [hunk 1]
```diff
@@ -57,23 +50,16 @@
...
-
-        this.agents = new HashMap<>();
-        findAgents();
-        for (Agent agent : agents) {
-            this.agents.put(agent.getId(), agent);
-        }
-        this.agentIds = ImmutableList.copyOf(this.agents.keySet());
+    public AgentRegistryImpl(Set<Agent> agents) throws IOException {
+        this.agents = new HashMap<>(agents.size());
+        agents.forEach(a -> this.agents.put(a.getId(), a));
```
Comment (garagatyi) [inline agents/che-core-api-agent/src/main/java/org/eclipse/che/api/agent/server/impl/AgentRegistryImpl.java]: Here it is possible that someone declared 2 equal agents. 
If all of them have equals and hashcode then multibinder will throw an error.
If multibinder is declared anywhere with support of duplicates then it won't throw exception.
If equals and hashcode is not overridden or overridden in a different ways then here duplicate agent will override previous one.

**Review by garagatyi — CHANGES_REQUESTED** [review]

File: agents/ls-csharp/src/main/java/org/eclipse/che/api/agent/LSCSharpAgent.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+/**
+ * Language server C# agent.
+ *
+ * @see Agent
+ *
+ * @author Anatolii Bazko
+ */
+@Singleton
+public class LSCSharpAgent implements Agent {
```
Comment (garagatyi) [inline agents/ls-csharp/src/main/java/org/eclipse/che/api/agent/LSCSharpAgent.java]: Looks like all the agents are very similar. Please consider creation of abstract base class for them.

**Review by garagatyi — APPROVED** [review]
Looks good. Great job!
Please consider improvements I commented.

**Comment by skabashnyuk** [comment]
ok

**Pull request merged**